### PR TITLE
Add support for emitting loop register annotations

### PR DESCRIPTION
**Pull request by boomanaiden154**
This patch adds support for emitting loop register annotations for performing benchmarking in loop mode. Tests are added in addition to new flags to control the functionality where appropriate.

### CODE REVIEW

**Comment by boomanaiden154** [comment]
Note that this is a stacked pull request on top of #51, so please don't merge it (if we get to that point) until #51 has already landed.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (ondrasej) [inline gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:289]: I'd put this into a library so that it can be unit tested. IMO unit tests are a much better fit for this piece of code, because it is complex enough, and we're testing directly the loop counter detection behavior, not any accidental other behaviors.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (ondrasej) [inline gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:288]: How often does this happen on the ComPile data set?

**Review by ondrasej — CHANGES_REQUESTED** [review]

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (boomanaiden154) [reply gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc]: Refactored out into a library (`basic_block_utils` from a dependency PR) and unit-tested.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (boomanaiden154) [reply gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:270]: I don't have any statistics on that currently. Will hopefully collect them once this script is working at scale.

File: gematria/datasets/basic_block_utils_test.cc
Comment (ondrasej) [inline gematria/datasets/basic_block_utils_test.cc:125]: This might break when LLVM changes register indices. That doesn't happen very often, but we've already been broken by this a couple of times.

A more reliable way would be to test that the returned value is one of the remaining GPRs.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (ondrasej) [reply gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:270]: No problem. But we should count them and then report the total number at the end.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (ondrasej) [reply gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc]: 👍

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (boomanaiden154) [reply gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:272]: Good point. The script now outputs information on this at the end.

File: gematria/datasets/basic_block_utils_test.cc
Comment (boomanaiden154) [reply gematria/datasets/basic_block_utils_test.cc]: Interesting. Didn't realize LLVM would change the register indices that often. Good point though that this test would be brittle regardless. I've changed it as you've suggested.

File: gematria/datasets/basic_block_utils.h
Comment (ondrasej) [inline gematria/datasets/basic_block_utils.h:37]: Please add a comment explaining what this does (returns a GPR that is not touched by the basic block). The name might sound like we're looking for some kind of counter used by the basic block.

It could also make sense to rename this to `getUnusedGPRegister()` to make it more clear what this does (instead of naming it by what it is used for).

**Review by ondrasej — CHANGES_REQUESTED** [review]

**Comment by ondrasej** [comment]
Please let me know once the parent PRs are merged and this is rebased, I'll review then.

**Comment by boomanaiden154** [comment]
> Please let me know once the parent PRs are merged and this is rebased, I'll review then.

Everything has been rebased now.

File: gematria/datasets/basic_block_utils.cc
Comment (ondrasej) [inline gematria/datasets/basic_block_utils.cc:131]: We should use the same check as in https://github.com/google/gematria/blob/ad44397422734b3e43765821b11842edc8c003e3/gematria/datasets/basic_block_utils.cc#L39 to make sure we're always moving towards the largest super register.

File: gematria/datasets/basic_block_utils.cc
Comment (ondrasej) [inline gematria/datasets/basic_block_utils.cc:157]: Shorter:
```cc
const MCInstrDesc& Desc = InstructionInfo.get(Instruction.mc_inst.getOpcode());
UsedRegisters.insert(Desc.implicit_uses().begin(), Desc.implicit_uses.end());
UsedRegisters.insert(Desc.implicit_defs().begin(), Desc.implicit_defs.end());
```

**Review by ondrasej — CHANGES_REQUESTED** [review]

**Pull request merged**
